src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,118 +248,118 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer, uint numBytes)
+        private unsafe void ParseEventBufferAndNotifyForEach(ReadOnlySpan<byte> buffer)
         {
-            Debug.Assert(buffer != null);
-            Debug.Assert(buffer.Length > 0);
-            Debug.Assert(numBytes <= buffer.Length);
-
-            numBytes = Math.Min(numBytes, (uint)buffer.Length);
+            ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
-            fixed (byte* b = buffer)
+            // Parse each event from the buffer and notify appropriate delegates
+            while (true)
             {
-                byte* pBufferLimit = b + numBytes;
-
-                Interop.Kernel32.FILE_NOTIFY_INFORMATION* info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
-
-                ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
+                // Validate the data we received in case it's corrupted. This can happen
+                // if we are watching files over the network with a poor connection.
 
-                // Parse each event from the buffer and notify appropriate delegates
-                do
+                // Verify the info object is within the expected bounds
+                if (sizeof(FILE_NOTIFY_INFORMATION) > (uint)buffer.Length)
                 {
-                    // Validate the data we received in case it's corrupted. This can happen
-                    // if we are watching files over the network with a poor connection.
+                    // Defensive check. We have not observed this corruption in practice.
+                    Debug.Assert(false);
+                    break;
+                }
+                ref readonly FILE_NOTIFY_INFORMATION info = ref MemoryMarshal.AsRef<FILE_NOTIFY_INFORMATION>(buffer);
 
-                    // Verify the info object is within the expected bounds
-                    if (info < b || (((byte*)info) + sizeof(Interop.Kernel32.FILE_NOTIFY_INFORMATION)) > pBufferLimit)
-                    {
+                if (info.FileNameLength > (uint)buffer.Length - sizeof(FILE_NOTIFY_INFORMATION))
+                {
+                    // We have observed this corruption in practice (https://github.com/dotnet/runtime/issues/40412)
```

Comment:
Why no assert here?

Reply:
My thinking was that we do not want to assert for invariants that we do not hold all the time. But I agree with you that we should have asserts here since the truncated data are most likely manifestation of an OS bug.